packages/jupyterlab-manager/src/semverregistry.ts

```diff
@@ -0,0 +1,147 @@
+ * that initial object registration can happen concurrent with the first
+ * requests.
+ *
+ */
+export class SemVerRegistry<T> {
+  /**
+   * Register a object with a given name and version.
+   */
+  register(options: SemVerRegistry.IRegistryData<T>): void {
+    const { name, version, object } = options;
+    let versions = this._registry.get(name);
+    if (versions === undefined) {
+      versions = new Map<string, T>();
+      this._registry.set(name, versions);
+    }
+
+    if (versions.has(version)) {
+      throw new Error(`Version ${version} of ${name} already registered.`);
+    }
+    versions.set(version, object);
+
+    // Some pending requests may now be satisfied, so schedule a retry.
+    if (this._pendingRequests.length > 0) {
+      this._throttle.invoke();
+    }
+  }
+
+  /**
+   * Get an object for a name and semVer range.
+   *
+   * @returns a promise resolving to the object with the maximum version
+   * satisfying the version range if one is registered, or undefined if one is
+   * not registered.
+   *
+   * #### Notes
+   * Resolution may be delayed for a short timeout to allow for some delay in
+   * registering the object.
+   */
+  async get(name: string, semVer: string): Promise<T | undefined> {
+    const mod = this._getSync(name, semVer);
+    if (mod) {
+      return mod;
+    }
+    const result = new PromiseDelegate<T | undefined>();
+    const timestamp = Date.now();
+    this._pendingRequests.push({ name, semVer, result, timestamp });
+    this._throttle.invoke();
+    return result.promise;
+  }
+
+  /**
+   * Synchronously get an object for a name and semVer range.
+   *
+   * @returns the object with the maximum version satisfying the version range
+   * or undefined if one does not exist.
+   */
+  private _getSync(name: string, semVer: string): T | undefined {
+    const versions = this._registry.get(name);
+    if (versions) {
+      const best = maxSatisfying(Array.from(versions.keys()), semVer);
+      if (best !== null) {
+        return versions.get(best);
+      }
+    }
+  }
+
+  /**
+   * Retry resolving pending requests.
+   */
+  private _retry(): void {
+    const now = Date.now();
+    // Try to resolve each pending request successfully.
+    this._pendingRequests.forEach(request => {
+      const { name: key, semVer: semver, result } = request;
+      const mod = this._getSync(key, semver);
+      if (mod) {
+        result.resolve(mod);
+        request.timestamp = null;
+      }
+    });
+
+    // Clear out all requests which were fulfilled or expired.
+    ArrayExt.removeAllWhere(
+      this._pendingRequests,
+      ({ result, timestamp }) => {
+        if (timestamp === null) {
+          // Request fulfilled, so remove the entry.
+          return true;
+        } else if (now - timestamp > TIMEOUT) {
+          // Request expired, so give up and remove the entry.
+          result.resolve(undefined);
+          return true;
```

Comment:
Does this logic mean that if the resolution takes more than `TIMEOUT` (300 ms) wall time, we give up?

Reply:
yep. 

Right now we basically immediately give up (no later loading of widgets supported).

Reply:
Just to make sure I understand, what problem does this PR resolve?

Reply:
Good question.

In general, the widget manager depends on the notebook, and custom widgets depend on the notebook manager, so we aren't guaranteed to have custom widgets loaded before a notebook is instantiated just from the dependency injection system. We noticed a problem when using @wolfv's new dynamic extension loader, which can load a custom widget like pythreejs or bqplot at app startup, but loads them asynchronously and in an unspecified order.

What this PR does is give extensions a chance to load custom widgets asynchronously near to when they are needed. We did pick kind of an arbitrary timeout to try to balance between being able to load something from the network and process it, and giving up being responsive about showing an error to the user.

Reply:
Sounds like a tricky problem. Let me see if I understand the timings:
- All extensions are activated (their promises pending).
- The layout restorer starts fetching its data.
- The notebook extension resolves its tracker.
- The widget manager registers a placeholder with the rendermime registry, and resolves its registry
- Extensions that supply widgets register with the registry
- The dynamic widget extension starts fetching remote scripts.
- The `app.restored` promise resolves.
- The settings registry starts fetching schemas to update settings of other extensions
- The layout restorer starts restoring trackers (e.g. notebooks)
- Widgets are instantiated
- Settings resolve for the dynamic widget extension, or its fetch results come back, and it now registers a widget type, too late.

Reply:
Would it make sense for the dynamic widget extension to not resolve its activate function before its network fetches resolve? It can potentially slow down the app start up, but it seems like the right place to wait for remote JS fetches. If so, the only place that could be a source of a race would be the settings fetch.

Reply:
Ah, I guess not, as I assume the remote URLs are also stored in settings..

Reply:
I don't think we can change the logic much but in the worst case scenario the user should see `Loading widget model...` for ~300ms max (which is blocking the notebook execution) but he should _only_ see it for the first time. I think that should be OK from a UX perspective, and buys us enough time to register the dynamically loaded widgets so restoration works as it should – otherwise simply re-executing the cell should do it.

Although that gives me an idea ... would it be possible to figure out wether a widget is being created from a restoration vs interactively? Because if yes, we should never wait interactively, and wait as long as possible in the restoration mode ... would that be possible cc @jasongrout